utils/src/ordered.rs

```diff
@@ -0,0 +1,1053 @@
+    }
+}
+
+impl<K: Ord + Clone, V: Clone> From<&[(K, V)]> for Map<K, V> {
+    fn from(items: &[(K, V)]) -> Self {
+        items.iter().cloned().collect()
+    }
+}
+
+impl<K: Ord, V, const N: usize> From<[(K, V); N]> for Map<K, V> {
+    fn from(items: [(K, V); N]) -> Self {
+        items.into_iter().collect()
+    }
+}
+
+impl<K: Ord + Clone, V: Clone, const N: usize> From<&[(K, V); N]> for Map<K, V> {
+    fn from(items: &[(K, V); N]) -> Self {
+        items.as_slice().into()
+    }
+}
+
+impl<K, V> From<Map<K, V>> for Vec<(K, V)> {
+    fn from(wrapped: Map<K, V>) -> Self {
+        wrapped.into_iter().collect()
+    }
+}
+
+impl<K: Write, V: Write> Write for Map<K, V> {
+    fn write(&self, buf: &mut impl BufMut) {
+        self.keys.write(buf);
+        self.values.write(buf);
+    }
+}
+
+impl<K: EncodeSize, V: EncodeSize> EncodeSize for Map<K, V> {
+    fn encode_size(&self) -> usize {
+        self.keys.encode_size() + self.values.encode_size()
+    }
+}
+
+impl<K: Read + Ord, V: Read> Read for Map<K, V> {
+    type Cfg = (RangeCfg<usize>, K::Cfg, V::Cfg);
+
+    fn read_cfg(buf: &mut impl Buf, cfg: &Self::Cfg) -> Result<Self, commonware_codec::Error> {
+        let (range_cfg, key_cfg, value_cfg) = cfg;
+        let keys = Set::<K>::read_cfg(buf, &(*range_cfg, key_cfg.clone()))?;
+        let values = Vec::<V>::read_cfg(buf, &(RangeCfg::exact(keys.len()), value_cfg.clone()))?;
+        Ok(Self { keys, values })
+    }
+}
+
+impl<K, V> IntoIterator for Map<K, V> {
+    type Item = (K, V);
+    type IntoIter = MapIntoIter<K, V>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        MapIntoIter {
+            keys: self.keys.into_iter(),
+            values: self.values.into_iter(),
+        }
+    }
+}
+
+impl<'a, K, V> IntoIterator for &'a Map<K, V> {
+    type Item = (&'a K, &'a V);
+    type IntoIter = core::iter::Zip<core::slice::Iter<'a, K>, core::slice::Iter<'a, V>>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.keys.iter().zip(self.values.iter())
+    }
+}
+
+/// An iterator over owned key-value pairs.
+pub struct MapIntoIter<K, V> {
+    keys: VecIntoIter<K>,
+    values: VecIntoIter<V>,
+}
+
+impl<K, V> Iterator for MapIntoIter<K, V> {
+    type Item = (K, V);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let key = self.keys.next()?;
+        let value = self.values.next()?;
+        Some((key, value))
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.keys.size_hint()
+    }
+}
+
+impl<K, V> ExactSizeIterator for MapIntoIter<K, V> {}
+
+impl<K, V> DoubleEndedIterator for MapIntoIter<K, V> {
+    fn next_back(&mut self) -> Option<Self::Item> {
+        let key = self.keys.next_back()?;
+        let value = self.values.next_back()?;
+        Some((key, value))
+    }
+}
+
+/// An ordered, deduplicated collection of key-value pairs with unique values.
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub struct BiMap<K, V> {
```

Comment:
Typically BiMap would have a function to get a key from a value, but maybe it's okay that this doesn't have that. Could consider `UniqueMap` if we don't have such a function... but it's probably alright as-is